compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp

```diff
@@ -370,15 +371,28 @@ class LLVMCPUTargetBackend final : public TargetBackend {
 
       std::string sourceFile = "";
       int sourceLine = 0;
+      std::string sourceContents = "";
       if (options.debugLevel >= 1) {
         if (auto loc = findFirstFileLoc(exportOp.getLoc())) {
           sourceFile = loc->getFilename().str();
           sourceLine = loc->getLine();
+          if (options.debugLevel >= 3) {
+            std::string error;
+            auto file = mlir::openInputFile(loc->getFilename(), &error);
```

Comment:
I'm not sure what this is doing, but I'm pretty sure it's not what we want. Not only will this load a file that won't always exist (remember that locs may reference files from way earlier in the pipeline) but the file may be binary, 100GB, etc.

What we want to embed is the `hal.executable.variant` contents.

Reply:
Here's an example of how to get the contents of an MLIR Operation into a string: https://github.com/openxla/iree/blob/09e5d1d9ad7d640b7b6078aab316ea0322c0f37f/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L235-L241

Pulling arbitrary source text could work _if_ the location is a `FileLineColLoc` (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td) and the file exists / is accessible. That could be interesting as a way to get portable .py sources embedded, for example.

The existing source information embedding works implicitly together with the location snapshotting and file writing that "dump executables" performs. This should decide if it wants to follow that or do something more explicit.

Reply:
For "something more explicit", we could have an enum flag (using `clEnumValN`) like `--iree-hal-executable-source-locations=[original, mlir]` (or other names, idk), that would let you choose between
* using the source locs already in there (what this PR does right now)
* using the MLIR `hal.executable.variant` contents

Decoupling from `--iree-hal-dump-executable-sources-to` would also make this easier to use.

Maybe for this PR start with just embedded MLIR contents, unless you really want input .py/.mlir files for your use cases. Once the plumbing is complete then we can figure out how to make it more flexible and what flags to define.

Reply:
@benvanik: agreed this is not desired, I was debugging here. 

The problem is at the `SerializeExecutablesPass`, the `hal.executable.variant` does not contain the correct source that is indicated by the `Loc` that is shown in tracy currently. If you compile a model (MobileBert for example), we'll either embed the source locs from the input mlir/py file (depending on input), or in the case of using `--iree-hal-dump-executable-sources-to`, the output from the `DumpExecutableSourcesPass` (what `mlir::openInputFile()` is reading as a test). 

At `SerializeExecutablesPass` this is no longer available. This gist shows `DumpExecutableSourcesPass` vs `SerializeExecutablesPass`: https://gist.github.com/KoolJBlack/cec5481510e3852d87e070f2dcc22aec

Essentially I need access to source higher up in the pipeline. With Scott's suggestion for an explicit flag: `--iree-hal-executable-source-locations=[original, mlir]`, I could leverage this to add the MLIR to the executable op at the `DumpExecutableSourcesPass` as an attribute and carry it down to `SerializeExecutablesPass`. This won't handle Locs that point to the input file however, and I'm not sure how to handle src locs that are provided with the input mlir that point to companion files (like python). This still per executable op however which could duplicate source until deduped during serialization.

Suggestions?

Reply:
> I could leverage this to add the MLIR to the executable op at the `DumpExecutableSourcesPass` as an attribute and carry it down to `SerializeExecutablesPass`.

Rather than overload what that pass does, I'd add a new pass (added to the pipeline at the same point here: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp). That would decouple this from "dump executable sources".

I was thinking start simple:

* If `--iree-hal-executable-debug-level=3` is set (or a new flag), have a pass there snapshot the variant op _as a string_ (not a file), stashed onto an attribute
* Later in executable serialization, look for that attribute and copy the text into what is serialized (for the runtime / Tracy to use)

Then after the first PR, consider also using MLIR source locations as a way to look up and embed input python files, if that would be useful (input python files can be referenced today but not portably)

Reply:
Just had a few more ideas, if snapshotting the variant as a string in an attribute is tricky (not sure how string escaping works there).

* Clone the variant ops into a module that is excluded from future compiler passes (not sure how excluding would work, but that would keep it in IR/memory the whole way, without worrying about serialize/deserializing)
* Dump the variant ops to _tool temp files_, separately from `DumpExecutableSourcesPass` (maybe in that pass, but at least to separate files), then remember the paths to those files in attributes, _not MLIR source locs_.